Added Voice assistance

drishtipeshwani opened the pull request:
- Added voice assistance functionality in our project. 
 - That is, other than typing the city name, we can also speak the city name and weather of that city will get displayed.
 
Fixed Issue #15

drishtipeshwani committed:
Initial changes

drishtipeshwani committed:
Initial changes

drishtipeshwani committed:
Added voice assistant functionality

drishtipeshwani committed:
Added voice assistant functionality

drishtipeshwani committed:
Fixed city value issue

drishtipeshwani committed:
Fixed city value issue

NoahCardoza commented:
I know you are still waiting on reviews from @DinuBotan and @shravani05 but some of my recent PR's have caused some merge conflicts.

Also, I wasn't able to get the speech to text to work. Every time I clicked the mic icon, the slash would reappear a second or two after clicking it without clicking it a second time... does that make sense?

I'm using the Brave Browser on macOS and no errors occurred in the console.

However, after testing in Chrome it seems to work. I'd suggest either debugging on the Brave Browser or disabling the feature when `!(navigator.brave && await navigator.brave.isBrave())`.

It also might be a good idea to set the search bar after updating the card based on the user's speech.

![image](https://user-images.githubusercontent.com/10343470/143145756-38d08ac1-02f5-4bb9-92b7-70f3467f3a95.png)

NoahCardoza commented on src/components/CurrentCityWeather/CurrentCityWeather.js:
```diff
@@ -1,23 +1,33 @@
-        "&units=metric" +
-        "&appid=" +
-        process.env.REACT_APP_APIKEY
+      city +
+      "&units=metric" +
+      "&appid=" +
+      process.env.REACT_APP_APIKEY
```
I saw this somewhere recently and have started using it since it makes thing a little easier to read:

```js
      "https://api.openweathermap.org/data/2.5/weather?"
      + `q=${city}&`
      + "units=metric&"
      + `appid=${process.env.REACT_APP_APIKEY}`
```

drishtipeshwani committed:
Merge branch 'main' into main

drishtipeshwani commented:
> I know you are still waiting on reviews from @DinuBotan and @shravani05 but some of my recent PR's have caused some merge conflicts.
> 
> Also, I wasn't able to get the speech to text to work. Every time I clicked the mic icon, the slash would reappear a second or two after clicking it without clicking it a second time... does that make sense?
> 
> I'm using the Brave Browser on macOS and no errors occurred in the console.
> 
> However, after testing in Chrome it seems to work. I'd suggest either debugging on the Brave Browser or disabling the feature when `!(navigator.brave && await navigator.brave.isBrave())`.
> 
> It also might be a good idea to set the search bar after updating the card based on the user's speech.
> 
> ![image](https://user-images.githubusercontent.com/10343470/143145756-38d08ac1-02f5-4bb9-92b7-70f3467f3a95.png)

Yaa, I should have taken care of supported browsers. These are the supported browsers - 
 - Chrome (desktop): this is by far the smoothest experience
 - Safari 14.1
 - Microsoft Edge
 - Chrome (Android): a word of warning about this platform, which is that there can be an annoying beeping sound when turning the microphone on. This is part of the Android OS and cannot be controlled from the browser
 - Android web view
 - Samsung Internet
 I have added a function that only displays the microphone icon if the browser is supported one. 
Regarding the displaying value in the search bar actually earlier we were searching weather based on city, but now we are searching based on coordinates so there were some issues regarding that that have been resolved in one of the pull request. I guess it would be good to merge that PR first and then I can make changes to prevent conflicts.

DinuBotan commented on src/components/CurrentCityWeather/CurrentCityWeather.js:
```diff
@@ -101,19 +111,42 @@ const CurrentCityWeather = () => {
 
+
+  useEffect(() => {
+    console.log(transcript)
```
This console log should probably be deleted.

drishtipeshwani replied:
Ya removed that

DinuBotan reviewed (commented):
Looks good and the voice assistance works nicely, but it only updates the location name and doesn't actually update the forecast for that location. I think that's because right now the forecast is fetched based on a city's coordinates and not on its name. 

<img width="760" alt="image" src="https://user-images.githubusercontent.com/61206601/143316374-81657883-9ad4-4eb6-981e-6b7515172211.png">

<img width="676" alt="Screenshot 2021-11-24 at 22 43 28" src="https://user-images.githubusercontent.com/61206601/143316515-0a95636e-b095-4a25-a4ef-1a64d857f9f4.png">

If so, then you would need to find a way to get the coordinates of the location that the user is saying or make requests to the API based on the city name instead of coordinates like we were doing initially (but this comes with a small problem that was solved in issue #21 ).

If you decide to go with the coordinates way and need some help, let me know.

drishtipeshwani committed:
Merge branch 'main' into main

drishtipeshwani committed:
Fetched data based on co-ordinates

drishtipeshwani commented:
> Looks good and the voice assistance works nicely, but it only updates the location name and doesn't actually update the forecast for that location. I think that's because right now the forecast is fetched based on a city's coordinates and not on its name.
> 
> <img alt="image" width="760" src="https://user-images.githubusercontent.com/61206601/143316374-81657883-9ad4-4eb6-981e-6b7515172211.png">
> 
> <img alt="Screenshot 2021-11-24 at 22 43 28" width="676" src="https://user-images.githubusercontent.com/61206601/143316515-0a95636e-b095-4a25-a4ef-1a64d857f9f4.png">
> 
> If so, then you would need to find a way to get the coordinates of the location that the user is saying or make requests to the API based on the city name instead of coordinates like we were doing initially (but this comes with a small problem that was solved in issue #21 ).
> 
> If you decide to go with the coordinates way and need some help, let me know.

Ya, I changed that. I added a function to get the coordinates of the city that we will speak and I am setting those coordinates. So able to fetch the correct data. However, I am not sure how to update the search bar because it is an Autocomplete component so you can set a default value at the beginning but not sure if we can change the value. Any idea about this?

drishtipeshwani commented:
Ya, I changed that. I added a function to get the coordinates of the city that we will speak and I am setting those coordinates. So able to fetch the correct data. However, I am not sure how to update the search bar because it is an Autocomplete component so you can set a default value at the beginning but not sure if we can change the value. Any idea about this?

drishtipeshwani commented:
However, the name of the place of the weather card is showing different, that issue has been fixed by Pratyush in the previous PR.

NoahCardoza commented:
> 

If you don't have any luck with this I can try looking for a solution tomorrow.

Also, the voice button still seems to show up in Brave (and not work). Maybe the little code snippet I gave you was bad?

drishtipeshwani commented:
Ya actually the npm library that I am using for this had a function to see if the current browser is supported or not, so I used that and it was working as Firefox is not a supported browser so it was not showing the microphone, not sure why it is not happening with Brave. I will put the condition you gave and see

drishtipeshwani commented:
The only thing remaining is to update the search bar with that place name, but I am not able to understand how to change the value as it is an Autocomplete component and has only a default value parameter.

drishtipeshwani committed:
Merge branch 'main' into main

drishtipeshwani committed:
Fixed browser support issue

drishtipeshwani committed:
Merge branch 'main' of github.com:drishtipeshwani/prep-project-4.1.4 into main

NoahCardoza commented on src/components/CurrentCityWeather/CurrentCityWeather.js:
```diff
@@ -100,19 +111,59 @@ const CurrentCityWeather = () => {
+            defaultValue={"New York, NY, USA"}
+            className="inputCity"
+          />
+          {browserSupportsSpeechRecognition && checkBraveBrowser ? (
```
You'll probably want to use `!` to make sure it's not Brave. Right now this code is making sure it is Brave.

Also, you'll have to be careful, `checkBraveBrowser` returns a `Promise` and a `Promise` is always true. However, you want to wait for the value returned by the promise before proceeding.

drishtipeshwani replied:
checkBraveBrowser is actually a function I have defined that will return a boolean value.

NoahCardoza replied:
Actually, `checkBraveBrowser` will return the `Promise` of a `Boolean` since you used the async/await.

![image](https://user-images.githubusercontent.com/10343470/143816654-032d3d06-4030-4a1f-8ec5-124099f20645.png)

As you can see it returns a Promise and since a Promise is an Object it's "true". Thus the function will always return "true" if it's not awaited or `.then`'ed.

NoahCardoza commented on src/components/CurrentCityWeather/CurrentCityWeather.js:
```diff
@@ -100,19 +111,59 @@ const CurrentCityWeather = () => {
+  }
+
+  async function checkBraveBrowser() {
+    if (navigator.brave && await navigator.brave.isBrave()) {
+      return false
+    } else {
+      return true
+    }
```
To check a boolean and return a boolean is a little bit redundant. Try this:

```suggestion
    return navigator.brave && await navigator.brave.isBrave();
```

drishtipeshwani replied:
Ya this way is better I will change to this

NoahCardoza reviewed (changes requested):
Promises can be a little tricky in React, if you need a hand let me know!

drishtipeshwani committed:
Merge branch 'main' into main

drishtipeshwani committed:
Shorten the checkBrowser function

Co-authored-by: Noah Cardoza <noahcardoza@gmail.com>

drishtipeshwani commented:
Does everything look good now ?
